external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala

```diff
@@ -0,0 +1,199 @@
+          } else None
+
+        // TODO: Support cases where topic+partitions are missing from one. Can happen in case of
+        // repartitioning.
+
+        case _ =>
+          None
+      }
+    }
+  }
+
+  override def toString(): String = {
+    offsets.toSeq.sortBy(_._1.topic).mkString("[", ", ", "]")
+  }
+}
+
+/** Companion object of the [[KafkaSourceOffset]] */
+private[kafka] object KafkaSourceOffset {
+
+  /** Returns [[KafkaSourceOffset]] from a Option[Offset]. */
+  def from(offsetOption: Option[Offset]): Option[KafkaSourceOffset] = {
+    offsetOption.map { offset =>
+      offset match {
+        case o: KafkaSourceOffset => o
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Invalid conversion from offset of ${offset.getClass} to $getClass")
+      }
+    }
+  }
+
+  /**
+   * Returns [[KafkaSourceOffset]] from a variable sequence of (topic, partitionId, offset)
+   * tuples.
+   */
+  def apply(data: (String, Int, Long)*): KafkaSourceOffset = {
+    val map = data.map { case (topic, partition, offset) =>
+        TopicAndPartition(topic, partition) -> offset }.toMap
+    KafkaSourceOffset(map)
+  }
+}
+
+
+/** A [[Source]] that reads data from Kafka */
+private[kafka] case class KafkaSource(
+    topics: Set[String],
+    params: Map[String, String])(implicit sqlContext: SQLContext) extends Source with Logging {
+
+  type OffsetMap = Map[TopicAndPartition, Long]
+
+  implicit private val encoder = ExpressionEncoder.tuple(
+    ExpressionEncoder[Array[Byte]](), ExpressionEncoder[Array[Byte]]())
+
+  @transient private val logicalPlan = StreamingRelation(this)
+  @transient private val kc = new KafkaCluster(params)
+  @transient private val topicAndPartitions = KafkaCluster.checkErrors(kc.getPartitions(topics))
+  @transient private[kafka] val initialOffsets = getInitialOffsets()
+
+  override def schema: StructType = encoder.schema
+
+  /** Returns the next batch of data that is available after `start`, if any is available. */
+  override def getNextBatch(start: Option[Offset]): Option[Batch] = {
+    val beginOffset: KafkaSourceOffset = KafkaSourceOffset.from(start).getOrElse(initialOffsets)
+    val latestOffset = getLatestOffsets()
+    logDebug(s"Latest offset: $latestOffset")
+
+    val offsetRanges = beginOffset to latestOffset
+    val kafkaParams = params
+    val encodingFunc = encoder.toRow _
+    val sparkContext = sqlContext.sparkContext
+
+    if (offsetRanges.nonEmpty) {
+      val rdd = KafkaUtils.createRDD[Array[Byte], Array[Byte], DefaultDecoder, DefaultDecoder](
+        sparkContext, kafkaParams, offsetRanges.toArray)
+      logInfo("Creating DF with offset ranges: " + offsetRanges)
+      Some(new Batch(latestOffset, sqlContext.createDataset(rdd).toDF))
+    } else {
+      None
+    }
+  }
+
+  def toDS(): Dataset[(Array[Byte], Array[Byte])] = {
+    toDF.as[(Array[Byte], Array[Byte])]
+  }
+
+  def toDF(): DataFrame = {
+    new DataFrame(sqlContext, logicalPlan)
+  }
```

Comment:
I would consider defining these in the tests since they are really only for testing.